From shaohua.li@intel.com  Fri Oct 26 22:23:02 2007
From: Shaohua Li <shaohua.li@intel.com>
Date: Wed, 24 Oct 2007 10:43:23 +0800
Subject: pcie port driver: correctly detect native PME feature
To: linux-pci <linux-pci@atrey.karlin.mff.cuni.cz>
Cc: Greg KH <greg@kroah.com>
Message-ID: <1193193803.27150.13.camel@sli10-conroe.sh.intel.com>


Native PME is capability of root port or root complex event collector.
It's not determined by PCI PME capability.

Signed-off-by: Shaohua Li <shaohua.li@intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/pci/pcie/portdrv_core.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

--- a/drivers/pci/pcie/portdrv_core.c
+++ b/drivers/pci/pcie/portdrv_core.c
@@ -192,9 +192,8 @@ static int get_port_device_capability(st
 		if (reg32 & SLOT_HP_CAPABLE_MASK)
 			services |= PCIE_PORT_SERVICE_HP;
 	} 
-	/* PME Capable */
-	pos = pci_find_capability(dev, PCI_CAP_ID_PME);
-	if (pos) 
+	/* PME Capable - root port capability */
+	if (((reg16 >> 4) & PORT_TYPE_MASK) == PCIE_RC_PORT)
 		services |= PCIE_PORT_SERVICE_PME;
 	
 	pos = PCI_CFG_SPACE_SIZE;
